docs/infra/develop-and-test-infrastructure-in-isolation-using-workspaces.md

```diff
@@ -0,0 +1,68 @@
+# Develop and test infrastructure in isolation using workspaces
+
+When developing infrastructure code, you often want to develop and test your changes in isolation so that:
+
+- Your changes do not impact other engineers on the team
+- Other engineers working on infrastructure do not revert your changes when making their own changes
+- Other engineers can review your changes before applying them
+
+This document describes a workflow that leverages [Terraform workspaces](https://developer.hashicorp.com/terraform/language/state/workspaces) for developing and testing infrastructure changes in isolation so that they can be tested and peer reviewed before being merged into main.
+
+## Overview
+
+By default, each Terraform root module has a single workspace named `default`. Workspaces allow you to deploy multiple instances of the root module configuration without configuring a new backend. When you run `terraform apply` in a separate workspace, a parallel set of infrastructure resources are created.
+
+There are a few notable differences with resources created in a non-default workspace:
+
+1. When using terraform, you cannot deploy the same resource with the same name. The `terraform apply` will fail with an error like "A resource with the ID already exists". Therefore, to avoid naming conflicts, we prefix the resource names with the workspace name.
+2. Some resources, such as database and storage buckets, enable deletion protection to prevent accidental deletion. However, non-default workspaces are intended to be temporary, so deletion protection is disabled in non-default workspaces.
+3. Resources that are difficult to create in isolation, such as [DNS records](https://github.com/navapbc/template-infra/blob/2cda6da18c84aa5a3dfb038ab32be4fac363af21/infra/modules/service/dns.tf#L3), are not created at all.
+
+## Development workflow
+
+Follow these steps if you want to develop and test a change to the service layer. Make the appropriate changes to the `-chdir` flag if you want to make a change to a different layer, such as the database layer or network layer.
+
+### 1. Create a new workspace
+
+First, make sure that the Terraform root module is initialized to the dev environment
+
+```bash
+terraform -chdir=infra/$app_name/service init -backend-config=dev.s3.tfbackend
```

Comment:
At some point, we should land on a documentation styleguide for consistency and approachability. I think all of the docs use `<APP_NAME>` to indicate fields that should be replaced by the person running the command.

Reply:
Our Makefile executes the following `init` for `infra-update-app-service`:

```sh
terraform -chdir="infra/$(APP_NAME)/service" init -input=false -reconfigure -backend-config="$(ENVIRONMENT).s3.tfbackend"
```

Would it be more consistent to include the `-input=false` and `-reconfigure` flags in this instruction? Are there potential issues the user might run into without these flags?

Reply:
Good point on style guide I'll update.

`-reconfigure` is worth including but `-input=false` is only for running in a CI context not interactively.